props/saber_BC_buttons.h

```diff
@@ -386,31 +1134,321 @@ EFFECT(volmax);     // for maximum volume reached
+    this->fadeout(SaberBase::sound_length);
+  }
+
+  void select() override {
+    cancelling_ = false;
+    mode::getSL<SPEC>()->SaySave();
+    SPEC::SteppedMode::select();
+  }
+
+  void exit() override {
+    PVLOG_NORMAL << "** Exit Volume Menu\n";
+    cancelling_ = true;
+    SPEC::MenuBase::exit();
+  }
+
+  bool mode_Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+    switch (EVENTID(button, event, 0)) {
+      // Custom button controls for BCVolumeMode
+      case EVENTID(BUTTON_POWER, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMaxVolume();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_SECOND_HELD_MEDIUM, 0):
+      case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMinVolume();
+        return true;
     }
+    // Use the select and exit controls from SelectCancelMode
+    return SPEC::SelectCancelMode::mode_Event2(button, event, modifiers);
+  }
+  bool cancelling_ = false;
+};
+
+#ifdef DYNAMIC_BLADE_LENGTH
+
+int selected_blade_;
+
+#ifdef BC_DUAL_BLADES
+template<class SPEC>
+struct BCSelectBladeMode : public SPEC::MenuBase {
+  int current_blade_ = 1; // Start with the first blade
+  ShowColorSingleBladeTemplate<Pulsing<ShowColorStyle,Black,800>,Pulsing<ShowColorStyle,Black,800>> highlighted_blade_;
+
+  uint16_t size() override {
+    return NUM_BLADES;
+  }
+  void mode_activate(bool onreturn) override {
+    mode::getSL<SPEC>()->SayEditBladeLength();
+    mode::getSL<SPEC>()->SaySelectBlade();
+    ShowColorStyle::SetColor(Color16(65535, 65535, 65535));
+    SPEC::SteppedMode::mode_activate(onreturn);
+    highlighted_blade_.Start(current_blade_);
+    PVLOG_NORMAL << "** Highlighting blade: " << current_blade_ << "\n";
+    say();
+  }
+
+  void next() override {
+    highlighted_blade_.Stop(current_blade_);
+    current_blade_ = (current_blade_ % NUM_BLADES) + 1;
+    highlighted_blade_.Start(current_blade_);
+    PVLOG_NORMAL << "** Highlighting blade: " << current_blade_ << "\n";
+    say();
+  }
+
+  void prev() override {
+    highlighted_blade_.Stop(current_blade_);
+    current_blade_ = (current_blade_ + NUM_BLADES - 2) % NUM_BLADES + 1;
+    highlighted_blade_.Start(current_blade_);
+    PVLOG_NORMAL << "** Highlighting blade: " << current_blade_ << "\n";
+    say();
+  }
+
+  void say() override {
+    mode::getSL<SPEC>()->SayBlade();
+    mode::getSL<SPEC>()->SayWhole(current_blade_);
+  }
+
+  void select() override {
+    // Set the current blade for the ChangeBladeLengthMode before transitioning
+    // Deactivate current mode and push to LED length adjustment mode for the selected blade
+    selected_blade_ = current_blade_;
+    highlighted_blade_.Stop(current_blade_);
+    SPEC::SteppedMode::mode_deactivate();
+    pushMode<typename SPEC::ChangeBladeLengthMode>();
+  }
+
+  void exit() override {
+    highlighted_blade_.Stop(current_blade_);
+    SPEC::MenuBase::exit();
+  }
+};
+#endif  // BC_DUAL_BLADES
+
+
+template<class SPEC>
```

Comment:
Besides handling different blades, is there any difference between this and the mode in modes/ ?

Reply:
Just the user experience mainly. I realize I probably don't need to override all the methods from the base if it inherits from ChangeBladeLengthBlade1, only select() and upadte() really, but I have a real hard time making that inheritance happen. Something about the naming, or circular dependency, or having the typedef use the same name etc... I got to a point it compiled and seemed to work, but ShowLengthStyle won't light up an LED.
I just left it as an independent thing once it worked.
If you prefer me to change it, great, but I'll need some coaching on what needs to be changed.

Reply:
> Just the user experience mainly.

That can mean anything though.

> If you prefer me to change it, great, but I'll need some coaching on what needs to be changed.

I don't mind the coaching, but I have to know what the intended differences are first.

Reply:
> I don't mind the coaching, but I have to know what the intended differences are first.



ok I'll explain instead of trying to figure it all out on my own (trying to save you hassle).
The only differences would be
steps_per_revolution 40.
Use SFX_volup instead of ccchange, plus delay the announce by 400ms.
SaySave on exit instead of SaySelect.

As far as what's confusing/not working when I try it: (this is my brain)
I see that BCChangeBladeLengthMode is mostly redundant of the already existing ChangeBladeLengthBlade1 and realize that's not the right way to do it likely.
I wanted to change it to just inherit what it could and only change what I wanted to customise. That's how it works, right?

Well I originally had the BCChangeBladeLengthMode using a "BC" version alias in the typedef as `typedef BCChangeBladeLengthMode<SPEC> BCChangeBladeLengthMode;`, but testing yielded no LED lighting up. So looking at ShowLengthStyle, I saw that it's looking for 'SPEC::ChangeBladeLengthMode>()->getLength();'.
(this is where I probably messed up) So instead of making another thing, and duplicating ShowLengthStyle to a new version like BCShowLengthStyle just so that it could use BCChangeBladeLengthMode... I just typedef-ed the BCChangeBladeLengthMode to the same alias name as the default ChangeBladeLengthBlade1 template uses, ChangeBladeLengthMode, thinking that would work since I wasn't using the actual ChangeBladeLengthBlade1 template in my spec, so the alias shouldn't be in conflict or anything.
Well that didn't work, I don't remember if it was a compiler error or the LED still wasn't lighting, but then I said let's do this right and use inheritance from the default version.
So I restored the BCversion alias in the typedef, and tried more things and it still didn't work.
At that point, i just left it as inheriting from MenuBse, fully redundantly overrode all the methods, and that's where it stands because it works.

Bottom line is I really only need to override steps_per_revolution, select(), and update().

Reply:
I indent to create a define which lets you specify which blades are "removable", and then change the default menu so that you can change the length of all the removable blades. When I do that, hopefully I can also make all of this simpler for you to use and override.

It's not always true that inheriting and overriding is the right thing to do. It kind of depends on how many things you want to change. If you want to change one thing, then inherit-and-override is almost certainly the right way to go, but if you want to change enough things that you have to override almost everything, then it might be better and easier to just copy-paste-modify, that is normal and reasonable.

I don't really have any major problems with this way of implementing it, other than that I kind of hope that I can change the existing implementation to make it easier for you to do what you want, and maybe then inherit-and-override will be the right way to solve things again.

Reply:
> I intend to create a define which lets you specify which blades are "removable", and then change the default menu so that you can change the length of all the removable blades. When I do that, hopefully I can also make all of this simpler for you to use and override.

Sounds good. Maybe it'll be what gets used here instead of NUM_BLADES.
So for now this can remain as-is? 
The only thing I'm not thrilled about is how it "cheats" to make use of ShowLengthStyle. It only works I think because I've aliased BCChangeBladeLengthMode as ChangeBladeLengthMode (the same thing the default spec uses.)
I think it really should be aliased as like BCChangeBladeLengthMenu instead, but then would need to also make a BCShowLengthStyle version too.
 Should I do that?

Reply:
This all has been resolved by inheriting from the SPEC directly instead of using the typedef-ed alias name.

Reply:
> > Bottom line is I really only need to override steps_per_revolution, select(), and update().
> 
> Can we try doing that? It seems like it should be nicer and smaller that way. (And I think other people may end up copying from your code, so I want it to be as neat at possible.)

Done.